core/src/main/scala/org/bitcoins/core/protocol/script/ScriptPubKey.scala

```diff
@@ -620,17 +623,67 @@ object ConditionalScriptPubKey extends ScriptFactory[ConditionalScriptPubKey] {
     fromAsm(asm)
   }
 
-  def isConditionalScriptPubKey(asm: Seq[ScriptToken]): Boolean = {
+  def isConditionalScriptPubKeyWithElseIndex(
+      asm: Seq[ScriptToken]): (Boolean, Int) = {
     val headIsOpIf = asm.headOption.contains(OP_IF)
-    lazy val containsOpElse = {
-      val opElseIndex = asm.indexOf(OP_ELSE)
-      opElseIndex != -1
+    lazy val endsWithEndIf = asm.last == OP_ENDIF
+
+    var opElseIndexOpt: Option[Int] = None
+
+    lazy val validConditionalTree = {
+      // Already validate that the first token is OP_IF, go to tail
+      // and start with depth 1 and no OP_ELSE found for that depth
+      val opElsePendingOpt = asm.zipWithIndex.tail
```

Comment:
This seems to be more readable IMO

```scala
      val opElsePendingOpt = asm.zipWithIndex.tail
        .foldLeft[Option[Vector[Boolean]]](Some(Vector(false))) {
          case (None, _) => // Invalid tree case, do no computation
            None
          case (Some(Vector()),_) =>
            // Case of additional asm after final OP_ENDIF
            None

          case (Some(opElseFoundAtDepth), (OP_IF | OP_NOTIF, _)) =>
            // Increase depth by one with OP_ELSE yet to be found for this depth
            Some(opElseFoundAtDepth :+ false)
          case (Some(opElseFoundAtDepth), (OP_ELSE, asmIndex)) =>
            if (opElseFoundAtDepth == Vector(false)) {
              // If first OP_ELSE at depth 1, set opElseIndex
              opElseIndexOpt = Some(asmIndex)
            }

            if (opElseFoundAtDepth.last) {
              // If OP_ELSE already found at this depth, invalid
              None
            } else {
              // Otherwise, set to found at this depth
              Some(
                opElseFoundAtDepth.updated(opElseFoundAtDepth.length - 1,
                  true))
            }

          case (Some(opElseFoundAtDepth), (OP_ENDIF, _)) =>
            if (opElseFoundAtDepth.last) {
              // If OP_ELSE found at this depth then valid, decrease depth by 1
              Some(opElseFoundAtDepth.dropRight(1))
            } else {
              // Otherwise, invalid
              None
            }
          case (Some(opElseFoundAtDepth), (_, _)) =>
            // Token not related to conditional structure, ignore
            Some(opElseFoundAtDepth)
        }
```